core/src/sumeragi/network_topology.rs

```diff
@@ -5,341 +5,183 @@
+        Topology {
+            sorted_peers: peers.into_iter().collect(),
+        }
     }
-
-    /// Answers if the consensus stage is required with the current number of peers.
+    /// Is consensus required, aka are there more than 1 peer.
     pub fn is_consensus_required(&self) -> bool {
         self.min_votes_for_commit() > 1
     }
-
-    /// The minimum number of signatures needed to commit a block
-    pub fn min_votes_for_commit(&self) -> usize {
-        2 * self.max_faults() + 1
-    }
-
-    /// The minimum number of signatures needed to perform a view change (change leader, proxy, etc.)
-    pub fn min_votes_for_view_change(&self) -> usize {
-        self.max_faults() + 1
-    }
-
-    /// Peers of set A. They participate in the consensus.
-    pub fn peers_set_a(&self) -> &[PeerId] {
-        let n_a_peers = 2 * self.max_faults() + 1;
-        &self.sorted_peers[..n_a_peers]
-    }
-
-    /// Peers of set B. The watch the consensus process.
-    pub fn peers_set_b(&self) -> &[PeerId] {
-        let n_a_peers = 2 * self.max_faults() + 1;
-        &self.sorted_peers[n_a_peers..]
-    }
-
-    /// The leader of the current round.
-    #[allow(clippy::expect_used)]
-    pub fn leader(&self) -> &PeerId {
-        self.peers_set_a()
-            .first()
-            .expect("Failed to get first peer.")
-    }
-
-    /// The proxy tail of the current round.
-    #[allow(clippy::expect_used)]
-    pub fn proxy_tail(&self) -> &PeerId {
-        self.peers_set_a().last().expect("Failed to get last peer.")
-    }
-
-    /// The peers that validate the block in discussion this round and vote for it to be accepted by the blockchain.
-    pub fn validating_peers(&self) -> &[PeerId] {
-        let a_set = self.peers_set_a();
-        if a_set.len() > 1 {
-            &a_set[1..(a_set.len() - 1)]
-        } else {
-            &[]
-        }
+    /// How many faulty peers can this topology tolerate.
+    pub fn max_faults(&self) -> usize {
+        (self.sorted_peers.len().saturating_sub(1)) / 3
     }
-
-    /// Get role of the peer by its id.
-    pub fn role(&self, peer_id: &PeerId) -> Role {
-        if self.leader() == peer_id {
-            Role::Leader
-        } else if self.proxy_tail() == peer_id {
-            Role::ProxyTail
-        } else if self.validating_peers().contains(peer_id) {
-            Role::ValidatingPeer
-        } else {
-            Role::ObservingPeer
-        }
+    /// The required amount of votes to commit a block with this topology.
+    pub fn min_votes_for_commit(&self) -> usize {
+        self.max_faults() * 2 + 1
     }
-
-    /// Verifies that this `message` was signed by the `signature` of a peer with specified `role`.
-    ///
-    /// # Errors
-    /// Fails if there are no such peer with this key and if signature verification fails
-    // TODO: Not used?
-    pub fn verify_signature_with_role(
+    /// Filter signatures by roles in the topology.
+    #[allow(clippy::comparison_chain)]
```

Comment:
```suggestion

```

Reply:
I'm keeping this one because the code is not better if I follow the lint suggestion.